### PR TITLE
Make CommandsCache play well with Windows

### PR DESCRIPTION
**Pull request by BYK**
Currently when you do `'git' in builtins.__xonsh_commands_cache__` on Windows you get `False` since the version stored in commands cache is `GIT.EXE`.

This patch amends its getter methods to respond to any version available on Windows. This fixes `curr_branch` not working at all on Windows platform since it only looks for `'git'` in command cache and bailds if its not there.

### CODE REVIEW

File: xonsh/environ.py [hunk 1]
```diff
@@ -823,31 +823,27 @@ def _yield_executables(directory, name):
...
+
+    cc = builtins.__xonsh_commands_cache__
+
     if ON_WINDOWS:
         # Windows users expect to be able to execute files in the same
         # directory without `./`
         cwd = _get_cwd()
         if os.path.isfile(name):
             return os.path.abspath(os.path.relpath(name, cwd))
+
```
Comment (scopatz) [inline xonsh/environ.py]: No need for all of this whitespace :)

File: xonsh/tools.py [hunk 2]
```diff
@@ -1560,8 +1560,18 @@ def __init__(self):
         self._alias_checksum = None
         self._path_mtime = -1
 
+    def __get_possible_names(self, key):
```
Comment (scopatz) [inline xonsh/tools.py]: Please don't start attrs or methods with double underscore because then they are name mangled by Python and live only on the class

Comment (scopatz) [inline xonsh/tools.py]: See https://docs.python.org/3.5/tutorial/classes.html#private-variables

> Any identifier of the form __spam (at least two leading underscores, at most one trailing underscore) is textually replaced with _classname__spam, where classname is the current class name with leading underscore(s) stripped. This mangling is done without regard to the syntactic position of the identifier, as long as it occurs within the definition of a class.

Comment (BYK) [inline xonsh/tools.py]: I know, this was my intention but then I decided I don't really need a private method :)

**Comment by scopatz** [comment]
Thanks for opening this @BYK!

File: xonsh/built_ins.py [hunk 3]
```diff
@@ -455,8 +455,8 @@ def run_subproc(cmds, captured=False):
         if (stdin is not None and
                 ENV.get('XONSH_STORE_STDIN') and
                 captured == 'object' and
-                'cat' in __xonsh_commands_cache__ and
-                'tee' in __xonsh_commands_cache__):
+                __xonsh_commands_cache__.locate_binary('cat') and
+                __xonsh_commands_cache__.locate_binary('tee')):
```
Comment (scopatz) [inline xonsh/built_ins.py]: These should be lazy, right?

File: xonsh/tools.py [hunk 4]
```diff
@@ -1560,6 +1560,17 @@ def __init__(self):
...
         self._path_mtime = -1
 
+    @staticmethod
+    def get_possible_names(key):
+        if ON_WINDOWS:
+            key = key.upper()
+            return {
+                key + ext
+                for ext in ([''] + builtins.__xonsh_env__['PATHEXT'])
+            }
```
Comment (scopatz) [inline xonsh/tools.py]: This should not be a set operation because PATHEXT is ordered by the windows spec.

File: xonsh/tools.py [hunk 5]
```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
...
+        """Locates an executable in the cache."""
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self._cmds_cache.keys()
+
+        if matches:
+            return self._cmds_cache[matches.pop()][0]
+        elif os.path.isfile(name) and name != os.path.basename(name):
+            return name
+
```
Comment (scopatz) [inline xonsh/tools.py]: Lots of extra vertical whitespace in these functions.

File: xonsh/tools.py [hunk 6]
```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
...
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self.all_commands.keys()
+
+        if matches:
+            return self.all_commands[matches.pop()][0]
+        elif os.path.isfile(name) and name != os.path.basename(name):
+            return name
+
+    def lazy_locate_binary(self, name):
+        """Locates an executable in the cache."""
```
Comment (scopatz) [inline xonsh/tools.py]: I think these two docstrings could be a little more descriptive and distinct

File: xonsh/tools.py [hunk 7]
```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
...
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self._cmds_cache.keys()
```
Comment (scopatz) [inline xonsh/tools.py]: This is a pretty inefficient way to check if a single name is in the cache.  Assuming possibilities returns a list or names, a faster implementation would be:

```python
match = None
for n in self.get_possible_names(name):
    if n in self._cmds_cache:
        match = n
        break
```

File: xonsh/tools.py [hunk 8]
```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
...
+    def locate_binary(self, name):
+        """Locates an executable on the file system."""
+
+        if ON_WINDOWS:
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+
+            local_bins = [full_name for full_name in self.get_possible_names(name)
+                          if os.path.isfile(full_name)]
```
Comment (scopatz) [inline xonsh/tools.py]: This should only have to find the first match that is a file, not test all of the possible matches for being a file each time.

File: xonsh/tools.py [hunk 9]
```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
...
+            if local_bins:
+                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self.all_commands.keys()
+
+        if matches:
+            return self.all_commands[matches.pop()][0]
```
Comment (scopatz) [inline xonsh/tools.py]: This function should only access `self.all_commands` once, to prevent excessive re-computation of the cache validity.

File: xonsh/built_ins.py (hunk 3 above)
Comment (scopatz) [inline xonsh/built_ins.py]: Or really, it maybe should be lazyin() thus actual path doen't matter here at all.  Or you could revert it to what it was.

Comment (BYK) [inline xonsh/built_ins.py]: Okay with `lazy` variant. Was mostly trying to keep it consistent with the current version: `in` is not lazy right?

File: xonsh/tools.py (hunk 4 above)
Comment (BYK) [inline xonsh/tools.py]: Oh, you mean this ordering actually conveys a preference! Makes sense, will change.

File: xonsh/tools.py (hunk 8 above)
Comment (BYK) [inline xonsh/tools.py]: Hah, I was thinking about generators and then ended up with a list. Will optimize :)

File: xonsh/tools.py (hunk 9 above)
Comment (BYK) [inline xonsh/tools.py]: Good catch, will fix.

File: xonsh/tools.py (hunk 6 above)
Comment (BYK) [inline xonsh/tools.py]: Sure.

File: xonsh/tools.py (hunk 7 above)
Comment (BYK) [inline xonsh/tools.py]: Why would it be inefficient? I'm assuming set intersection does the same. Building a new set and using `pop` on it is rather inefficient though, yes.

File: xonsh/tools.py (hunk 5 above)
Comment (BYK) [inline xonsh/tools.py]: I find new lines making functions easier to follow by sectioning them. Will try to keep them minimal but TBH as an outsider, I find most of Xonsh code hard to follow due to this reason.

File: xonsh/tools.py (hunk 7 above)
Comment (scopatz) [inline xonsh/tools.py]: Yeah, it is the building of the set of keys that takes up the time, coupled with the fact that dictionary lookup is one of the fastest things that you can do in Python  Here is some simple timeit:

```
scopatz@artemis ~ $ timeit "next((x for x in range(1000) if x == 1), None)"
1000000 loops, best of 3: 600 ns per loop
scopatz@artemis ~ $ d = {i: i for i in range(1000)}
scopatz@artemis ~ $ p = {1}
scopatz@artemis ~ $ timeit "p & d.keys()"
100000 loops, best of 3: 11.5 µs per loop
scopatz@artemis ~ $ 11.5e3 / 600
19.166666666666668
scopatz@artemis ~ $ timeit "next((x for x in d if x == 1), None)"
1000000 loops, best of 3: 392 ns per loop
scopatz@artemis ~ $ 11.5e3 / 392
29.336734693877553
```

So this is about ~20x - 30x slower with a len-1000 sized cache.  Note that my cache is about 3600.

File: xonsh/tools.py [hunk 10]
```diff
@@ -1572,6 +1574,20 @@ def __len__(self):
     def __getitem__(self, key):
         return self.all_commands[key]
 
+    def is_empty(self):
```
Comment (scopatz) [inline xonsh/tools.py]: This needs a docstring.

File: xonsh/tools.py (hunk 5 above)
Comment (scopatz) [inline xonsh/tools.py]: I find them hugely distracting and harder to understand algorithms because less code fits on a page :). Also, PEP8 basically says to leave them out.  I think that this is mostly a style thing, so I defer to PEP8

Comment (melund) [inline xonsh/tools.py]: So do I. If you need to section different areas to make it easier to understand, you can do it with a comment line.

**Comment by melund** [comment]
Thanks  for tackling this @BYK. I will give it a thorough look when you push again.

**Comment by BYK** [comment]
@melund @scopatz - Ready for the next round. I really don't agree with you guys' "no empty lines" thing but I'll avoid them as much as possible since this is 1 against 2 :D (also yay consistency!)

File: xonsh/tools.py [hunk 11]
```diff
@@ -1648,34 +1650,36 @@ def lazyget(self, key, default=None):
...
-                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
 
-        possibilities = self.get_possible_names(name)
-        matches = possibilities & self.all_commands.keys()
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
```
Comment (melund) [inline xonsh/tools.py]: Should it not be enough to call `os.path.abspath(local_bin)`. What does the `relpath` do?

Comment (BYK) [inline xonsh/tools.py]: Don't have an opinion or preference on this. Just copied what the old code did :)

File: xonsh/tools.py [hunk 12]
```diff
@@ -1648,34 +1650,36 @@ def lazyget(self, key, default=None):
...
 
-        if matches:
-            return self.all_commands[matches.pop()][0]
+        cmds = self.all_commands
+        cached = next((cmd for cmd in possibilities if cmd in cmds), None)
```
Comment (melund) [inline xonsh/tools.py]: Since you just accessed `self.all_commands` I think you can use the lazy methods here: 
`cached = next((cmd for cmd in possibilities if cmd in self._cmds_cache), None)`

File: xonsh/tools.py [hunk 13]
```diff
@@ -1648,34 +1650,36 @@ def lazyget(self, key, default=None):
...
+        if cached:
+            return cmds[cached][0]
```
Comment (melund) [inline xonsh/tools.py]: Same here: `return self._cmds_cache[cached][0]`

File: xonsh/tools.py (hunk 12 above)
Comment (BYK) [inline xonsh/tools.py]: Ah, very good point. Will do.

Comment (BYK) [inline xonsh/tools.py]: Well, on second thought not sure if that'd provide any benefits. And it would make the code look a bit more complex. What do you think?

File: xonsh/tools.py [hunk 14]
```diff
@@ -1648,34 +1650,36 @@ def lazyget(self, key, default=None):
...
         elif os.path.isfile(name) and name != os.path.basename(name):
             return name
 
     def lazy_locate_binary(self, name):
```
Comment (melund) [inline xonsh/tools.py]: Should `lazy_locate_binary` not also look in the current directory, similar to what `locate_binary` does. I they share so much code that you could potentially call `lazy_locate_binary()` from `locate_binary()`. With the addition that `locate_binary()` should just update the command cache first.

Comment (BYK) [inline xonsh/tools.py]: >Should lazy_locate_binary not also look in the current directory, similar to what locate_binary does.

I'm not so sure about that. I feel like `lazy_` methods should be very fast and directory look up is not fast. It's also not using the cache so I decided not to add that logic here.

>I think they share so much code that you could potentially call lazy_locate_binary() from locate_binary(). With the addition that locate_binary() should just update the command cache first. 

This also occurred to me but I didn't spend too much time to find a good way to merge them. If you feel strongly I'll think harder ;)

**Comment by melund** [comment]
Argghh. Did I just comment on something that was outdated. Hmm. I will just redo my comments. Sorry :)

**Comment by BYK** [comment]
>Argghh. Did I just comment on something that was outdated. Hmm. I will just redo my comments. Sorry :)

Not sure if the places you commented on were outdated and they were certainly useful :)

**Comment by melund** [comment]
It is the same code. I was just looking at old commits, so GitHub hides the comments.

File: xonsh/tools.py (hunk 12 above)
Comment (melund) [inline xonsh/tools.py]: When ever you call the non-lazy methods, it will recompute all the hashes to see if it is still valid. So you only want to do that one time. 
`cmds = self.all_commands`

Afterwards you can just use the lazy methods.

File: xonsh/tools.py [hunk 15]
```diff
@@ -1634,13 +1654,45 @@ def lazyget(self, key, default=None):
...
+            # directory without `./`
+            cwd = _get_cwd()
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
+
+        cmds = self.all_commands
```
Comment (melund) [inline xonsh/tools.py]: Basically, you only want to access `all_commands` property ones. Afterwards use the lazy methods.

File: xonsh/tools.py [hunk 16]
```diff
@@ -1634,13 +1654,45 @@ def lazyget(self, key, default=None):
         """A lazy value getter."""
         return self._cmds_cache.get(key, default)
 
+    def locate_binary(self, name):
```
Comment (melund) [inline xonsh/tools.py]: If you move the stuff about the current directory to `lazy_locate_binary()`  and then here you just do

```
def locate_binary(self, name):
    cmd = self.all_commands
    return self.lazy_locate_binary(name)
```

Comment (BYK) [inline xonsh/tools.py]: Good idea :)

File: xonsh/commands_cache.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
+# -*- coding: utf-8 -*-
+import builtins
+import os
+from collections import abc as abc
```
Comment (scopatz) [inline xonsh/commands_cache.py]: This should be `import collections.abc as abc` to allow this to be a lazy import

**Comment by scopatz** [comment]
Also, `xonsh.commands_cache` needs a new `docs/api/commands_cache.rst` file.

File: xonsh/commands_cache.py (hunk 17 above)
Comment (BYK) [inline xonsh/commands_cache.py]: Ah, so far with PyCharm's automatic extraction feature :)

**Comment by scopatz** [comment]
Thanks a million @BYK!

File: xonsh/commands_cache.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+        if ON_WINDOWS:
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
```
Comment (melund) [inline xonsh/commands_cache.py]: I know this is already merged. But I think could be simplified a bit. Especially if @scopatz want to leave out the `_get_cwd()` bit. `os.path.abspath()` already works relative to current directory. So we could just do:
```python 
return os.path.abspath(local_bin)
```

Comment (scopatz) [inline xonsh/commands_cache.py]: Yeah, please feel free to clean this up!
